main/reference/vstorage-ref.md

```diff
@@ -1,13 +1,83 @@
-# VStorage Reference
+# VStorage in Agoric
 
-See also:
+In the Agoric platform, VStorage, or a Virtual Storage, is a key-value store that:
 
-- [Querying VStorage](/guides/getting-started/contract-rpc#querying-vstorage)
-- [Publishing to chainStorage](../guides/zoe/pub-to-storage)
-- [x/vstorage module](https://github.com/Agoric/agoric-sdk/tree/003f0c2232815a8d64a3f9a5b05521a10160ce34/golang/cosmos/x/vstorage#readme)
+- provides a **read-only interface for clients of the consensus layer**. Clients can query the stored data using specific paths.
+- is organized in a **hierarchical, path-based structure**. Each fragment of data is stored at a specific path, and clients can query these paths to retrieve the data.
+- can be written through a specialized API called **[chainStorage](https://docs.agoric.com/guides/zoe/pub-to-storage.html#publishing-to-chainstorage)**. This API allows the VM to manage data storage in a controlled and secure manner.
+
+![vstorage query diagram](../guides/getting-started/assets/vstorage-brand-q.svg)
+
+## VStorage Hierarchy
+
+VStorage is structured as a tree with paths and nodes that store the actual data. This design facilitates easy querying while ensuring data consistency and security.
+
+For example, here is an example of [Inter Protocol key structure](https://github.com/Agoric/agoric-sdk/tree/agoric-upgrade-13/packages/inter-protocol#reading-data-off-chain) inside Agoric:
+
+```
+- published
+    - vaultFactory
+        - governance
+        - metrics
+        - managers
+            - manager0
+                - metrics
+                - governance
+                - vaults
+                    - vault0
+```
+
+You can then use this hierarchy to access the data fragments.
+```sh
+# lists vaults
+$ agd query vstorage keys 'published.vaultFactory.managers.manager0.vaults'
+children:
+- vault0
+```
+
+## VStorage with CLI
+VStorage can be accessed via CLI using:
+
+```sh
+$ agd /[--node {url}] query vstorage {path}
+```
+
+For example:
+```sh
+$ agd query vstorage path published.agoricNames
+children:
+- brand
+- installation
+- instance
+...
+```
+
+<!-- mention the constant marshalling in follow --> 
+With Agoric CLI, you can also use `follow` command to support vstorage query plus some of the marshalling conventions discussed below:
+```sh
+$ agoric follow -lF :published.agoricNames.brand
+[
+  [
+    "BLD",
+    slotToVal("board0566","Alleged: BLD brand"),
+  ],
+  [
+    "IST",
+    slotToVal("board0257","Alleged: IST brand"),
+  ],
+...
+]
+```
+
+## VStorage Query API
+
+## VStorage chainStorage API
 
-## vstorage: top level keys
+## VStorage Viewer
```

Comment:
These headings are taken from an semi-automated test. see the commit message for https://github.com/Agoric/documentation/commit/5607e3bf139bbe90aefd93b227e48a53a5a096e7 for details

I didn't manage to put the test into ci in such a way that changing the heading here without changing the test would be flagged as an inconsistency. I wonder what it would take to do that.

Reply:
1. I will revert these changes. I understand. 
2. Maybe run a script that opens the MD file and imports the JSON object with headings from the test and compares the two. If the headings are dependent on the result of a test, we can add it as a subsequent step in the CI.